Script user input improvement

pmiossec opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

## Proposed changes

In scripts command line arguments, the user is allows to:
* specify multiple user inputs
* specify a label as a reminder of what is requested (that will appear in the title of the input popup)
* specify or build a default value (that the user could modify before validating)

Example: `bundle create {{UserInput:Bundle filename|{cLocalBranch}.bundle|}} -{UserInput:Nb commits||} {cLocalBranch}`

2 inputs will be requested.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

<!-- TODO -->

### After

![image](https://github.com/gitextensions/gitextensions/assets/460196/cdadf6c0-ae93-4271-a6ff-25dd071da28c)

## Test methodology <!-- How did you ensure quality? -->

- Unit tests
- Manual

## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 33.33.33
- Build d735064e042defdc43432bcf68246cc392cc165f (Dirty)
- Git 2.42.0.windows.2
- Microsoft Windows NT 10.0.22621.0
- .NET 6.0.21
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```



## Merge strategy

<!-- Change the following if the merge strategy should be changed:
- Squash merge (maintainer to decide merge message, PR submitter should cleanup commits/messages at PR approval).
- Rebase merge (PR submitter must change the commit message for the last commit).
- Merge commit. (PR submitter to rebase and squash before merges).
- To be decided later.
The maintainer may still request the contributor to squash and rebase, to make sure that merges and commit messages are clarified.
-->

I agree that the maintainer **Rebase** merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented on GitUI/ScriptsEngine/ScriptOptionsParser.cs:
```diff
@@ -106,6 +109,23 @@ public static (string? arguments, bool abort) Parse(string? arguments, IGitUICom
+            Match match;
+            while ((match = userInputRegex.Match(arguments)).Success)
+            {
+                var defaultValue = Parse(match.Groups["defaultValue"].Value, uiCommands, owner, scriptHostControl, simplePromptCreator, filePromptCreator);
```
var

mstv commented on GitUI/ScriptsEngine/FormFilePrompt.cs:
```diff
@@ -64,4 +64,17 @@ private void Translate()
         }
     }
+
+    public class FilePromptCreator : IFilePromptCreator
+    {
+        public IUserInputPrompt Create()
+        {
+            return new FormFilePrompt();
+        }
+    }
+
+    public interface IFilePromptCreator
+    {
+        IUserInputPrompt Create();
+    }
```
I doubt that @RussKie will like this as it is.
❔
- Interface first
- Move to separate file `FilePromptCreator.cs`.
- Register IFilePromptCreator in Service Registry
- Usage: `uiCommands.GetRequiredService<IFilePromptCreator>().Create()`

RussKie replied:
Indeed :)

RussKie replied:
The concrete implementation should also remain non-public, as it should be reachable only via the interface registration.

mstv commented on GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs:
```diff
@@ -20,6 +20,7 @@ public partial class ScriptsSettingsPage : SettingsPageWithHeader
 
 User Input:
 {UserInput}
+{UserInput:a popup label|a default value|}
```
❓ more intuitive and less error-prone (missing trailing "|"):
```suggestion
{UserInput}
{UserInput:label}
{UserInput:label=default value}
```

RussKie replied:
{UserInput:label=defaultValue}

pmiossec replied:
We can find a better syntax like maybe `{UserInput[label=default value]}` but I really like to keep a end character otherwise it makes the parsing a hell if you want to have the possibility to make the default value a parsable expression with variables (like the example given in the description where the filename default value is built using the branch name variable)

mstv replied:
> it makes the parsing a hell

implement once, use many times
Does it really matter whether the default value is terminated by "}" or "|"? If only the interior of "{...}" is parsed, the default value end matches "$".

mstv reviewed (commented):
As already mentioned, multiple use of one input value is a very useful feature - e,g. "command {UserInput:A=a} ... {UserInput:B=b} ... {UserInput:A} ...".
Omitted label, i.e. empty string, automatically brings downwards compatibility.

RussKie commented on GitUI/ScriptsEngine/IUserInputPrompt.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace GitUI.ScriptsEngine
+{
+    public interface IUserInputPrompt : IDisposable
+    {
+        DialogResult ShowDialog(IWin32Window owner);
+        string UserInput { get; }
+    }
```
Please provide xml-docs

pmiossec commented:
I think I fixed all the points raised. Changes undo a lot of changes so I have rewritten the history (sorry) so that it's clearer now.

@mstv "multiple use of one input value" is now working as you expect.

@mstv  & @RussKie I tried to implement the syntax `{UserInput:label=default value}` you both proposed.
The RegEx is now a lot more complex to understand but it seems to work reasonably well...

mstv commented on GitUI/ScriptsEngine/FilePromptCreator.cs:
```diff
@@ -0,0 +1,15 @@
+﻿namespace GitUI.ScriptsEngine
+{
```
```suggestion
﻿namespace GitUI.ScriptsEngine;

```

mstv commented on GitUI/ScriptsEngine/IUserInputPrompt.cs:
```diff
@@ -0,0 +1,18 @@
+﻿namespace GitUI.ScriptsEngine
+{
```
```suggestion
﻿namespace GitUI.ScriptsEngine;

```

mstv commented on GitUI/ScriptsEngine/ScriptOptionsParser.cs:
```diff
@@ -106,6 +110,27 @@ public static (string? arguments, bool abort) Parse(string? arguments, IGitUICom
+                {
+                    prompt.ShowDialog(owner);
+                    arguments = ReplaceOption($"UserInput:{label}", arguments, prompt.UserInput);
+                    arguments = ReplaceOption(match.Value.Substring(1, match.Value.Length - 2), arguments, prompt.UserInput);
```
❔ Should we support quoting by double braces, too? Or is it handled by `ReplaceOption` already?

pmiossec replied:
It is handled inside `ReplaceOption()`.

mstv commented on GitUI/ScriptsEngine/SimplePromptCreator.cs:
```diff
@@ -0,0 +1,15 @@
+﻿namespace GitUI.ScriptsEngine
+{
```
```suggestion
﻿namespace GitUI.ScriptsEngine;

```

mstv commented on UnitTests/GitUI.Tests/Script/ScriptOptionsParserTests.cs:
```diff
@@ -368,13 +368,130 @@ public void ParseScriptArguments_resolve_QuotedWithBackslashAtEnd()
+        [Test]
+        public void Parse_should_parse_userInput_and_replace_all_with_same_label()
+        {
+            _commands.GetService<ISimplePromptCreator>().Returns(new SimplePromptCreatorForTest(("input label1", "value", "foo1")));
```
```suggestion
            _commands.GetService<ISimplePromptCreator>().ReturnsOnce(new SimplePromptCreatorForTest(("input label1", "value", "foo1")));
```
in order to verify that the dialog pops up only once - if NSubstitute supports it.

pmiossec replied:
Not possible but checked by line 453 (where we check that the input mock is displayed only the number of timed configured).

mstv reviewed (commented):
LGTM in a quick review, have not run yet

mstv commented on GitUI/ScriptsEngine/FormFilePrompt.cs:
```diff
@@ -4,9 +4,9 @@
 namespace GitUI.ScriptsEngine
 {
-    public partial class FormFilePrompt : GitExtensionsForm
+    public partial class FormFilePrompt : GitExtensionsForm, IUserInputPrompt
```
```suggestion
    internal partial class FormFilePrompt : GitExtensionsForm, IUserInputPrompt
```

mstv commented on GitUI/ScriptsEngine/SimplePrompt.cs:
```diff
@@ -1,12 +1,16 @@
+namespace GitUI.ScriptsEngine
 {
-    public partial class SimplePrompt : Form
+    public partial class SimplePrompt : Form, IUserInputPrompt
```
```suggestion
    internal partial class SimplePrompt : Form, IUserInputPrompt
```

mstv commented on GitUI/ScriptsEngine/SimplePrompt.cs:
```diff
@@ -1,12 +1,16 @@
         {
             InitializeComponent();
+            txt_UserInput.Text = defaultValue;
+            Text = label;
```
```suggestion
            if (!string.IsNullOrWhitespace(label))
            {
                Text = label;
            }
```

pmiossec replied:
The RegEx guarantees that there will be at least 1 character but why not if the user want to shoot in its own feet... (fixed)

RussKie commented on GitUI/ScriptsEngine/SimplePromptCreator.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace GitUI.ScriptsEngine;
+
+public interface ISimplePromptCreator
```
Does it have to be public?

pmiossec replied:
No. fixed.

mstv reviewed (commented):
The parser seems to work well. 👍

Cancellation of the input dialog should throw a (silent) `OperationAbortedException`.
`Esc` should deselect the selection if any, else cancel the input.

The label should be used as label (perhaps with ":" appended), but not as caption.
The caption should rather be `User input for script {name}`.

pmiossec commented:
> The parser seems to work well. 👍
> 
> Cancellation of the input dialog should throw a (silent) `OperationAbortedException`. `Esc` should deselect the selection if any, else cancel the input.
> 
> The label should be used as label (perhaps with ":" appended), but not as caption. The caption should rather be `User input for script {name}`.

Done.

mstv commented on GitUI/ScriptsEngine/ScriptsManager.ScriptRunner.cs:
```diff
@@ -55,98 +55,106 @@ private static bool RunScriptInternal(ScriptInfo script, IWin32Window owner, IGi
-                return true;
+                    return true;
+                }
+                catch (OperationCanceledException)
+                {
+                    MessageBox.Show(owner, TranslatedStrings.ScriptUserCanceledRun, TranslatedStrings.ScriptText, MessageBoxButtons.OK, MessageBoxIcon.Information);
+                    return false;
+                }
```
Exceptions are already handled by the single caller `RunScript`, i.e. move it there, please.
```suggestion
                catch (OperationCanceledException)
                {
                    MessageBox.Show(owner, TranslatedStrings.ScriptUserCanceledRun, script.Name, MessageBoxButtons.OK, MessageBoxIcon.None);
                    return false;
                }
```

mstv reviewed (commented):
final nits

RussKie commented on GitUI/ScriptsEngine/ScriptOptionsParser.cs:
```diff
@@ -106,6 +110,37 @@ public static (string? arguments, bool abort) Parse(string? arguments, IGitUICom
+                    DialogResult result = prompt.ShowDialog(owner);
+                    if (result != DialogResult.OK)
+                    {
+                        throw new OperationCanceledException();
```
😨 Why are we throwing here?
It looks as though we're using exceptions for the flow control, which is a code smell.
Throwing exceptions is quite expensive, and they represent an exceptional state, which user cancelling the dialog isn't.

pmiossec replied:
>Throwing exceptions is quite expensive

Performance doesn't matter here because the exception is thrown only one time to display a user popup.

>and they represent an exceptional state, which user cancelling the dialog isn't.

I think the cancelling of the dialog is exceptional otherwise we would have already an issue opened because the error popup is absolutely not good 🤣 

I really think a refactoring here is overkill as the method is returning at the moment only a value and we will have to refactor it to return also a state enum (Success/Abort/UserAbort), handle well this state for all arguments and handle it in the calling code. I predict more than 100 lines of new code and changes just to handle the case of 2 arguments `UserFile` and `UserInput` that ask inputs from user and that a bad/no input should abort the script. That's the handling of these 2 arguments and the possibility to abort which is "exceptional" 😉

I really don't want to go this way....

RussKie commented on GitUI/ScriptsEngine/ScriptOptionsParser.cs:
```diff
@@ -459,24 +494,30 @@ private static string GetRemotePath(string url)
+                        DialogResult result = prompt.ShowDialog(owner);
+                        if (result == DialogResult.Cancel)
+                        {
+                            throw new OperationCanceledException();
```
ditto

RussKie commented on GitUI/ScriptsEngine/ScriptsManager.ScriptRunner.cs:
```diff
@@ -55,98 +55,106 @@ private static bool RunScriptInternal(ScriptInfo script, IWin32Window owner, IGi
-                return true;
+                    return true;
+                }
+                catch (OperationCanceledException)
+                {
+                    MessageBox.Show(owner, TranslatedStrings.ScriptUserCanceledRun, TranslatedStrings.ScriptText, MessageBoxButtons.OK, MessageBoxIcon.Information);
+                    return false;
+                }
```
Why do we need this at all?

pmiossec replied:
Because otherwise the user have an frightening error message dialog that absolutely doesn't correspond to the reason why we abort the launch of the script:

![image](https://github.com/gitextensions/gitextensions/assets/460196/ab67a954-4437-4e96-97af-9d54f9be343a)

RussKie replied:
To me this is a confirmation that we're doing it incorrectly :)
I think this is a misplaced responsibility, please have a look at https://github.com/gitextensions/gitextensions/commit/4b772c4729a6dc6299f9bdebaa69fe3edf7e60a0

pmiossec replied:
@RussKie Not really convinced but OK for me. Have a look at other changes required by this solution and provide feedback before I spend time to adapt tests (because with this solution, it's now less straightforward and I don't want to spend too much time on it if later adaptations are needed).

pmiossec replied:
@RussKie Friendly reminder 😉

RussKie commented on GitUI/ScriptsEngine/SimplePrompt.Designer.cs:
```diff
@@ -46,41 +47,42 @@ private void InitializeComponent()
 
         #endregion
 
         private Button btn_OK;
         private TextBox txt_UserInput;
```
Please rename these and adjust the handles names
```suggestion
        private Button btnOk;
        private TextBox txtUserInput;
```

RussKie commented:
Apologies, the rl has taken priority. I'll try to look into it early next
week.

On Fri, 17 Nov 2023, 9:09 pm Philippe Miossec, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitUI/ScriptsEngine/ScriptsManager.ScriptRunner.cs
> <https://github.com/gitextensions/gitextensions/pull/11281#discussion_r1397027125>
> :
>
> > +                catch (OperationCanceledException)
> +                {
> +                    MessageBox.Show(owner, TranslatedStrings.ScriptUserCanceledRun, TranslatedStrings.ScriptText, MessageBoxButtons.OK, MessageBoxIcon.Information);
> +                    return false;
> +                }
>
> @RussKie <https://github.com/RussKie> Friendly reminder 😉
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/11281#discussion_r1397027125>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXTS42PN7CUZZ27MG43YE4ZUFAVCNFSM6AAAAAA6D3XM7OVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMZWGUZTGOBYGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RussKie commented on GitUI/ScriptsEngine/ScriptsManager.ScriptRunner.cs:
```diff
@@ -27,6 +32,70 @@ public static bool RunScript(ScriptInfo script, IWin32Window owner, IGitUIComman
                 }
             }
 
+            private static (string? arguments, bool abort, bool cancel) ParseUserInputs(string scriptName, string? arguments, IGitUICommands uiCommands, IWin32Window owner, IScriptOptionsProvider? scriptOptionsProvider = null)
```
Do we have tests for this?

mstv replied:
Yes, but they fail at the moment because they need to call `ParseUserInputs` instead of `ScriptOptionsParser.Parse`.

RussKie reviewed (approved):
👍 modulo the question about the tests

mstv reviewed (commented):
just a little cleanup yet

mstv commented on GitUI/ScriptsEngine/ScriptOptionsParser.cs:
```diff
@@ -283,12 +281,12 @@ private static string GetRemotePath(string url)
-            in IList<IGitRef> currentRemoteBranches, GitRevision currentRevision, string currentRemote)
+        private static string? ParseScriptArguments(string arguments, string option, IWin32Window owner, IScriptOptionsProvider? scriptOptionsProvider,
+            IGitUICommands uiCommands, IReadOnlyList<GitRevision> allSelectedRevisions, in IList<IGitRef> selectedTags,
+            in IList<IGitRef> selectedBranches, in IList<IGitRef> selectedLocalBranches, in IList<IGitRef> selectedRemoteBranches,
+            in IList<string> selectedRemotes, GitRevision selectedRevision, in IList<IGitRef> currentTags,
+            in IList<IGitRef> currentBranches, in IList<IGitRef> currentLocalBranches, in IList<IGitRef> currentRemoteBranches,
+            GitRevision currentRevision, string currentRemote)
```
This is effectively the same code. The change does not improve the readability.
If you really want to change it, then one argument per line.

RussKie replied:
I still think we should use a struct or a class here.

pmiossec replied:
Reverted (like all the below changes). Sorry, I have used the refactoring feature of VS without seeing that it changed all the arguments formatting.
I have squashed everything to remove these changes made in one commit and reverted in a later one and now history's much cleaner.

mstv commented on GitUI/ScriptsEngine/ScriptOptionsParser.cs:
```diff
@@ -572,14 +553,14 @@ public string CreateOption(string option, bool quoted)
+                string currentRemote)
+                => ScriptOptionsParser.ParseScriptArguments(arguments, option, owner, scriptOptionsProvider, uiCommands, allSelectedRevisions, selectedTags,
+                    selectedBranches, selectedLocalBranches, selectedRemoteBranches, selectedRemotes, selectedRevision, currentTags,
+                    currentBranches, currentLocalBranches, currentRemoteBranches, currentRevision, currentRemote);
```
ditto

mstv commented on UnitTests/GitUI.Tests/Script/ScriptOptionsParserTests.cs:
```diff
@@ -197,11 +197,11 @@ public void Parse_should_parse_s_arguments()
-                selectedBranches: null, selectedLocalBranches: null, selectedRemoteBranches: null, selectedRemotes: null, selectedRevision: null,
-                currentTags: null,
-                currentBranches: null, currentLocalBranches: null, currentRemoteBranches: null, currentRevision: null, currentRemote: null);
+                arguments: "{openUrl} https://gitlab.com{cDefaultRemotePathFromUrl}/tree/{sBranch}", option: "cDefaultRemotePathFromUrl", owner: null,
+                scriptOptionsProvider: null, uiCommands: null, allSelectedRevisions: null, selectedTags: null, selectedBranches: null,
+                selectedLocalBranches: null, selectedRemoteBranches: null, selectedRemotes: null, selectedRevision: null, currentTags: null,
+                currentBranches: null,
+                currentLocalBranches: null, currentRemoteBranches: null, currentRevision: null, currentRemote: null);
```
ditto + many more occurrences

mstv commented on UnitTests/GitUI.Tests/Script/ScriptOptionsParserTests.cs:
```diff
@@ -362,7 +362,7 @@ public void ParseScriptArguments_resolve_HEAD_with_checkedout_branch_name()
 
             string result = ScriptOptionsParser.GetTestAccessor().ParseScriptArguments(
-                arguments: "{" + option + "}", option,
+                arguments: "{" + option + "}", option: option,
```
If the compiler does not force us to add this, we should not add that redundancy.

mstv commented on UnitTests/GitUI.Tests/Script/ScriptOptionsParserTests.cs:
```diff
@@ -379,7 +379,7 @@ public void ParseScriptArguments_resolve_HEAD_with_detached_head_hash()
 
             string result = ScriptOptionsParser.GetTestAccessor().ParseScriptArguments(
-                arguments: "{" + option + "}", option,
+                arguments: "{" + option + "}", option: option,
```
ditto

mstv commented on UnitTests/GitUI.Tests/Script/ScriptOptionsParserTests.cs:
```diff
@@ -147,7 +147,7 @@ public void Parse_should_return_unmodified_arguments_if_no_options_matched()
             const string arguments = "{openUrl} https://gitlab.com{zeDefaultRemotePathFromUrl}/tree/{zeBranch}";
 
-            (string? arguments, bool abort) result = ScriptOptionsParser.Parse(arguments: arguments, _commands, owner: null, scriptOptionsProvider: null);
+            (string? arguments, bool abort) result = ScriptOptionsParser.Parse(arguments: arguments, uiCommands: _commands, owner: null, scriptOptionsProvider: null);
```
ditto

mstv commented on UnitTests/GitUI.Tests/Script/ScriptOptionsParserTests.cs:
```diff
@@ -422,7 +422,7 @@ public void Parse_should_resolve_IScriptOptionsProvider_options()
             _scriptOptionsProvider.GetValue(option2).Returns(value2);
 
-            (string? arguments, bool abort) = ScriptOptionsParser.Parse("foo {{" + option1 + "}} bar {" + option2 + "}", Substitute.For<IGitUICommands>(), owner: null, _scriptOptionsProvider);
+            (string? arguments, bool abort) = ScriptOptionsParser.Parse("foo {{" + option1 + "}} bar {" + option2 + "}", Substitute.For<IGitUICommands>(), owner: null, scriptOptionsProvider: _scriptOptionsProvider);
```
ditto

mstv commented on UnitTests/GitUI.Tests/Script/ScriptsManagerScriptRunnerTests.cs:
```diff
@@ -0,0 +1,141 @@
+using NSubstitute;
+using static GitUI.ScriptsEngine.ScriptsManager;
+
+namespace GitUITests.Script
+{
```
```suggestion
namespace GitUITests.Script;

```

mstv reviewed (commented):
nits
